---
date: "2019-08-05T14:07:44+02:00"
title: Template Walk-through
weight: 8
---


```{r setup, include=FALSE, results='hide', message=FALSE, warning=FALSE}
# hide all code chunks in the output, but show errors
knitr::opts_chunk$set(echo = TRUE,       # hide all code chunks in output
                      error = TRUE,       # show errors if they appear, but don't stop
                      fig.width = 6*1.25, # Figure width
                      fig.height = 6      # Figure height
                     )
# set default NA to - in output, define figure width/height
options(knitr.kable.NA = "-")

# Installing required packages for this template
required_packages <- c("knitr",       # create output docs
                       "dplyr",       # clean/shape data
                       "forcats",     # clean/shape data
                       "ggplot2",     # create plots and charts
                       "sitrep",      # MSF field epi functions
                       "linelist",    # Functions for cleaning/standardising data
                       "incidence",   # create epicurves
                       "aweek",       # define epi weeks
                       "epitools",    # 2x2 tables and other epi goodness
                       "sf",          # encode spatial vector data
                       "ggspatial")   # plot maps

for (pkg in required_packages) {
  # install packages if not already present
  if (!pkg %in% rownames(installed.packages())) {
    install.packages(pkg)
  }
  
  # load packages to this current session 
  library(pkg, character.only = TRUE)
}


# set default text size to 16 for plots
# give classic black/white axes for plots
ggplot2::theme_set(theme_classic(base_size = 18))

# Set the day that defines the beginning of your epiweek.
# you can start the week on any day of the week
# (the ISO standard is to start on Monday) 
aweek::set_week_start("Monday")
```

## Overview of this page
This page will demonstrate use of the Acute Jaundice Syndrome (AJS) outbreak template using data from an outbreak in Am Timan, Chad. This first page covers the initial data importation and cleaning steps in the template.  

**Opening the correct template**
[instructions here on opening the correct template]

**Access the datasets for this walk-through**
[instructions and link to dataset here]

**Access the final RMarkdown script used in this walk-through**  
[link to download Chad AJS RMarkdown here]

**RMarkdown resources**  
Review the page "Reading a RMarkdown Script", or reference these resources: [Rmarkdown introduction](https://rmarkdown.rstudio.com/articles_intro.html) and [Rmarkdown authoring basics](https://rmarkdown.rstudio.com/authoring_basics.html).  

  


### AJS/HEV outbreak in Am Timan, Chad (2016-2017)

We will be using data and examples from a real outbreak of Hepatitis E virus (HEV) infection which occurred in the Chadian town of Am Timan between October 2016 and April 2017  

<img src="/images/AJS_Article.png" alt="AJS Article snip", width = "40%">

[Link to publication](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0188240)  

*DISCLAIMER: The real data from the outbreak has been used for the training materials linked to the outbreak templates of the R4epis project. The Chadian Ministry of Health (MoH) has approved the use of this data for training purposes. Please note, that some data has been adapted in order to best achieve training objectives. Also, the GPS coordinates included in the dataset do not correspond to real cases identified during this outbreak. They have been generated exclusively for training purposes.*  

In early September 2016, a cluster of severely ill pregnant women with jaundice was detected at the Am Timan Hospital in the maternity ward. Following rapid tests conducted in Am Timan and confirmatory testing in the Netherlands, this was confirmed as due to HEV infection. Thus the MoH and MSF outbreak response was activated. The response consisted of four components: 1) active community based surveillance (CBS), 2) clinical assessment and management of ill cases at the hospital, 3) water chlorination activities at most water points in the town, 4) hygiene promotion.  

The outbreak linelist data was combined data from the active CBS data and the clinical data from the hospital. The CBS functioned with community health workers visiting all households in the town every two weeks and actively searching for people with Acute Jaundice Syndrome (AJS). For this group of persons (suspected cases), only those that were visibly ill or that were pregnant were referred to the hospital for clinical assessment and admission if required. Persons that self-reported to the hospital or that arrived after referral would undergo a clinical assessment and a rapid test for HEV for diagnosis. Thus only for people assessed at the hospital were we able to capture a confirmed case status.  

For the duration of the outbreak we detected 1193 suspected cases, 100 confirmed cases and discarded 150 cases with AJS who were not positive for HEV infection.  

## Template header and introduction

The very top of the template consists of a header surrounded by `---` lines. Here you may want to edit the title of your document. The other settings in the header define the default document type produced (Microsoft Word) when the RMarkdown is "knit".

The "Introduction to this template" section is for your information only and this text can be deleted before knitting the document.  

<img src="/images/AJS_Intro.png" alt="AJS Header and Introduction snip", width = "75%">


## Package installation
This section begins with text which explains the purpose of the following code chunk. **For your final report you may delete the text, but you must keep this code chunk!** The code will run without producing any output in your report, but is required to load packages necessary to produce the report.

The code chunk does the following:  

* Set default settings for the RMarkdown such as figure size, and handling of errors  

* Set the default way of displaying missing values in tables as "-"  

* Create a vector of names of packages required for the template to run. Then each package is installed if it is not already installed, and each package is loaded for use in the current R session.

* Set default plot theme text size and epidemiological weeks are set to begin on Mondays.  

## define current week 

```{r}
reporting_week <- "2017-W52"  # Set the reporting week
```

## Importing the dataset

**In this template section you choose the appropriate code chunk to importing your dataset.** 

<img src="/images/AJS_Data_Import.png" alt="Data import options snip", width = "75%">

For this exercise our dataset is a **non-standardized Excel file** from MSF's response an outbreak of AJS in Am Timan, Chad, so we use the `read_nonDHIS_data` code chunk, specifically the line that uses the import() function of the rio package to import an Excel file that is not password protected.  

Note that writing sometimes errors can occur if you provide an incorrect file name or path.


```{r eval=FALSE}
## Read data ------------------------------------
# Excel file
# to read in a specific sheet use "which"
linelist_raw <- rio::import("AJS_AmTiman.xlsx", which = "linelist")
```



```{r include=FALSE, results='hide', message=FALSE, warning=FALSE}
# This code chunk runs in the background and actually imports the dataset while transforming "." to NA.
# The import code chunk above (with eval=FALSE) does not run, but shows the viewer the default import code.
linelist_raw <- rio::import("AJS_AmTiman.xlsx", which = "linelist", na=".")
```


## Aligning the dataset with the data dictionary  
When using a non-standardized dataset, the variable names and values must be aligned with the data dictionary. This can take some time and requires looking back-and-forth between your dataset, RStudio, and the MSF data dictionary. In this example, here are the steps taken:  

### Review the MSF data dictionary
Use this command (found in the template) to view the MSF data dictionary for the disease ("AJS" in this example)

```{r eval = FALSE}
# creates object linelist_dict using the msf_dict function from the sitrep package
linelist_dict <- msf_dict("AJS", compact = FALSE) %>%
  select(option_code, option_name, everything())
```

You can view the data dictionary by running the command `View(linelist_dict)`, or by clicking the name linelist_dict in the RStudio Environment (upper-right RStudio pane)  

![](/images/AJS_Data_Dictionary.png)

### Clean variable names  
Uncomment the appropriate lines of code to make a copy of the dataset, naming the copy "linelist_cleaned". 

```{r}
## make a copy of your orginal dataset and name it linelist_cleaned
linelist_cleaned <- linelist_raw
```

Use the clean_labels function from the epitrix package to fix any variable names with non-standard syntax.
```{r}
# this function is preset rules for variable naming 
# for example it changes spaces and dots to "_" and characters to lowercase
cleaned_colnames <- epitrix::clean_labels(colnames(linelist_raw))

# overwrite variable names with defined clean names
colnames(linelist_cleaned) <- cleaned_colnames
```

### Re-name variables to align with data dictionary  

Standardized variable names are required for this template to work smoothly, and the variable names in our dataset do not align with the names this template expects. The template offers code that you can uncomment if you only have to change a few variable names (see below). **However, in this example we need to change many variable names. MSF developed a special function to help map our variables to the expected variables.**  

* **Paste the disease-specific rename helper into the script.** This command is found in the template. For our example we run the command `msf_dict_rename_helper("AJS")` and a block of code text is copied to the clipboard. Return to the template script and paste from the clipboard into the template. For the AJS template, the code that is pasted will look like the code below. This code uses the function rename() to change variable names.

<img src="/images/AJS_Rename_Variables_Empty.png" alt="rename variables empty snip", width = "75%">

* **Complete the mapping of variable names.** To the right of the equals sign, and before the comma, type the *exact* names of the variables from your dataset that correspond to the expected MSF data dictionary variables on the left. *If there is a data dictionary variable that is not present in your dataset, be sure to comment (#) out that line, as shown in the GIF below.*

![](/images/AJS_Variable_Names.gif)  

**Tip: **If you see this error:  
`Error in is_symbol(expr) : argument "expr" is missing, with no default`  
then you likely forgot to comment (#) the line for a variable you did not use.  

And now you can see the result:  

<img src="/images/AJS_Rename_Variables_Proof.png" alt="rename variables proof", width = "85%">


```{r include=FALSE, results='hide', message=FALSE, warning=FALSE}
## Add the appropriate column names after the equals signs

linelist_cleaned <- rename(linelist_cleaned,
#  event_file_type                   =   , # TEXT                # (Not in dataset)
  case_number                       =   hevid, # TEXT
  date_of_admission                 =   dtmedical, # DATE
  detected_by                       =   referredby, # TEXT
  patient_facility_type             =   hospitalised, # TEXT
#  msf_involvement                   =   , # TEXT                # (Not in dataset)
  age_years                         =   age, # INTEGER_POSITIVE
#  age_months                        =   , # INTEGER_POSITIVE   # (Dataset only has years)
#  age_days                          =   , # INTEGER_POSITIVE   # (Dataset only has years)
  sex                               =   sex, # TEXT
  patient_origin                    =   quartier, # ORG UNIT
#  residential_status_brief          =   , # TEXT               # (Not in dataset)
#  arrival_date_in_area_if_3m        =   , # DATE               # (Not in dataset)
#  ever_received_hepatitis_e_vaccine =   , # TEXT               # (Not in dataset)
#  date_of_last_vaccination          =   , # DATE               # (Not in dataset)
  date_of_onset                     =   dtjaundice, # DATE
#  history_of_fever                  =   , # BOOLEAN            # (Not in dataset)
  fever                             =   medfever, # BOOLEAN
  nausea_anorexia                   =   mednausea, # BOOLEAN
  vomiting                          =   medvomit, # BOOLEAN
  epigastric_pain_heartburn         =   medepigastric, # BOOLEAN
  generalized_itch                  =   meditching, # BOOLEAN
  headache                          =   medheadache, # BOOLEAN
  joint_pains                       =   medarthralgia, # BOOLEAN
  diarrhoea                         =   meddiar, # BOOLEAN
  bleeding                          =   medbleeding, # BOOLEAN
#  convulsions                       =   , # BOOLEAN            # (Not in dataset)
  mental_state                      =   medmental, # BOOLEAN    # !(Needs value cleaning)
  other_symptoms                    =   medother, # BOOLEAN
  other_cases_in_hh                 =   medothhhajs, # BOOLEAN
#  traditional_medicines             =   , # BOOLEAN            # (Not in dataset)
#  traditional_medicine_details      =   , # TEXT               # (Not in dataset)
#  recent_travel                     =   , # BOOLEAN            # (Not in dataset)
#  water_source                      =   , # TEXT               # !(Split across many variables)
  malaria_rdt_at_admission          =   medmalrdt, # TEXT
  hep_b_rdt                         =   medhepb, # TEXT
  hep_c_rdt                         =   medhepc, # TEXT
  hep_e_rdt                         =   medhevrdt, # TEXT
#  dengue_rdt                        =   , # TEXT               # (Not in dataset)
  date_lab_sample_taken             =   medblooddt, # DATE
  test_hepatitis_a                  =   medhavelisa, # TEXT
  test_hepatitis_b                  =   medhbvelisa, # TEXT
  test_hepatitis_c                  =   medhcvelisa, # TEXT
  test_hepatitis_e_virus            =   medhevelisa, # TEXT
  test_hepatitis_e_igm              =   hevrecent, # TEXT
#  test_hepatitis_e_igg              =   , # TEXT               # (In same variable as elisa result)
  test_hepatitis_e_genotype         =   hevgenotype, # TEXT
#  malaria_blood_film                =   , # TEXT               # (Not in dataset)
  dengue                            =   dengue, # TEXT
  yellow_fever                      =   yf, # TEXT
#  lassa_fever                       =   , # TEXT               # (Not in dataset)
#  typhoid                           =   , # TEXT               # (Not in dataset)
  chikungunya_onyongnyong           =   chik, # TEXT
#  ebola_marburg                     =   , # TEXT               # (Not in dataset)
  other_arthropod_transmitted_virus =   arbovpcr, # TEXT
#  other_pathogen                    =   , # TEXT               # (Not in dataset)
#  lab_comments                      =   , # TEXT               # (Not in dataset)
  pregnant                          =   medpreg, # TEXT
  trimester                         =   medpregtri, # TEXT
#  foetus_alive_at_admission         =   , # TEXT               # (Not in dataset)
#  delivery_event                    =   , # TRUE_ONLY          # (Not in dataset)
  pregnancy_outcome_at_exit         =   medppoutcome, # TEXT
  exit_status                       =   outcomehp, # TEXT
#  date_of_exit                      =   , # DATE               # (Not in dataset)
#  time_to_death                     =   , # TEXT               # (Not in dataset)
  treatment_facility_site           =   hpid, # TEXT
#  treatment_location                =   , # ORGANISATION_UNIT  # (Not in dataset)
  patient_origin_free_text          =   block # TEXT
)
```


## Provide population counts  
In order to calculate attack and mortality rates by region and age groups, the population counts for these categories are collected. If you choose not to complete this section, please delete or comment (#) the default gen_population() code and expect that later sections on attack and mortality rates will give errors instead of output. 

**The template allows several ways to provide population figures so that they are stored in data frames with specific variable names.** Of course, the specific age groupings and region names in your data frames may differ from the examples.  

You can do one of the following:  
1) Enter the counts directly **[need clarification here as this is not yet built]**  
2) Use the function gen_population() to derive the counts from proportions  
3) Import a spreadsheet file with the counts in the correct format  

![](/images/AJS_Population_Counts.png)  

For the AJS example, we can use the population Excel spreadsheet, importing it with the template code as below. Note that the variables names have been edited to match the expected names ("patient_origin" and "population").  

>The file path you provide to the function may differ. A few notes about file paths:  
* If copying a file path from Microsoft computer, you will need to reverse any slashes from \ to /.  
* [NOTES ON USING HERE and RIO::IMPORT]  

```{r}
population_data_region <- rio::import("AJS_AmTiman_population_revised_for_template.xlsx", which="quartier")
```

## Lab data
[Paragraph about lab data here - how it is used]  



## Creating and cleaning variables  
The variables and values in our example AJS dataset are still very far from the standard AJS MSF data dictionary. In the data cleaning step we will need to do the following:  

### **Re-code missing values from periods (.) to `NA` at import**  

Currently, missing values in the dataset are represented by a period (.). This causes problems because R expects missing values to be coded as `NA`. For example, if we attempt to set the variable date_of_onset to class Date, R gives an error because the periods are considered character values and do not fit the expected YYYY-MM-DD date format.  

To fix this problem we scroll up to the import section of the template and insert an additional argument in the import command. The argument `na = "."` specifies a period as the value in the Excel sheet that R will consider to be "missing". As the data are imported, all cells with that value are changed to `NA`.  

**Remember that after making this change to the import command you must re-run the entire script. This way, the dataset is imported correctly AND the changes to variable names, etc. are also performed on this newly imported dataset** 

```{r eval=FALSE}
# The old import command
linelist_raw <- rio::import("AJS_AmTiman.xlsx", which = "linelist")

# The revised import command (don't forget the comma between arguments!)
linelist_raw <- rio::import("AJS_AmTiman.xlsx", which = "linelist", na = ".")
```



### **Set date variables to class Date**
In linelist_cleaned there are seven variables containing dates (dateentry, epiquestdate, date_of_admission, date_of_onset, date_lab_sample_taken, dthospitalisation, dtdeath).  

**IF** your dataset *completely* aligns with the MSF data dictionary then use the code in the template that converts all the date variables to class Date at once.

```{r eval=FALSE}
# Use this is your dataset aligns with the data dictionary
# Creates list of date variables from the data dictionary
DATEVARS <- filter(linelist_dict, data_element_valuetype == "DATE") %>% 
  select(data_element_shortname) # select date vars
# change to the date variables found in your dataset to class date 
linelist_cleaned <- linelist_cleaned %>%
  mutate_at(DATEVARS$data_element_shortname, as.Date)
```


However, if your dataset is at all unstandard you will need to use this code instead.  
```{r}
# Use this if your dataset does not align with the data dictionary
  linelist_cleaned <- linelist_cleaned %>%
    mutate_at(vars(matches("date|Date")), linelist::guess_dates,
            error_tolerance = 0.5)
```

In our example dataset, this code converted many of our date variables, but did not detect the variables named with "dt", such as "dtdeath" and "dthospitalization". We must convert those manually:
```{r}
# Individually convert other date columns
class(linelist_cleaned$dthospitalisation)

linelist_cleaned$dthospitalisation <- linelist::guess_dates(linelist_cleaned$dthospitalisation)

class(linelist_cleaned$dthospitalisation)

# Repeat for variable "dtdeath"
linelist_cleaned$dtdeath <- linelist::guess_dates(linelist_cleaned$dtdeath)
```

### **Re-code values in patient_facility_type**  

When assigning variables to match the data dictionary, we used the variable "hospitalisation" as the variable "patient_facility_type". In the data dictionary, patient_facility_type should have values of "Inpatient" or "Outpatient." Currently, the values are "Oui" and "Non".  

This variable is referenced in later chunks, in which the analyses are restricted to observations where patient_facility_type == "Inpatient". Thus, we need to change the values to match the data dictionary and what the later code expects.  

We can modify a chunk from the cleaning section of the template. The code below re-defines linelist_cleaned as itself  mutated such that in the case when it's value equals "Oui", the value is changed (~) to "Inpatient", and so on. If applicable, write the lines from most specific at top to most general at bottom (see the section below on creating a case definition variable).

As you modify this chunk for your own situation, remember to include commas at the end of each case_when() line (except the last one), and that each left-hand side (LHS) of the ~ must be a logical statement, not just a value. You can read more about case_when() [here](https://dplyr.tidyverse.org/reference/case_when.html).  

**Tip: All the right-hand side (RHS) values must be the same class (either character, numeric, etc.). So, if your RHS values are character and you want one RHS value to be missing, you cannot just write `NA` on the RHS. Instead you must use the special character version of `NA` : `NA_character_`.** 

```{r}
# Check the values in patient_facility_type
table(linelist_cleaned$patient_facility_type, useNA = "always")

# Convert the values
linelist_cleaned <- linelist_cleaned %>% 
  mutate(patient_facility_type = case_when(
    patient_facility_type == "Oui" ~ "Inpatient",
    patient_facility_type == "Non" ~ "Outpatient"
  ))

# Re-check the values in patient_facility_type
table(linelist_cleaned$patient_facility_type, useNA = "always")

```

```{r}
# create an epiweek variable
# floor_day shortens to only give you the week number (rather than including day as well) 
# factor includes all weeks between the min and max as levels (useful for zero count weeks)
linelist_cleaned$epiweek <- aweek::date2week(linelist_cleaned$date_of_onset, 
                                             floor_day = TRUE, 
                                             factor = TRUE)
```


We ignore the section on "number of days under observation"" because our dataset does not have a date of exit variable.

Skip the optional age group variables because only have age_years

```{r}
## create an age group variable by specifying categorical breaks
linelist_cleaned$age_group <- age_categories(linelist_cleaned$age_years, 
                                             breakers = c(0, 3, 15, 30, 45))
```

### Create DIED variable  
When performing this step, we recognize that our current values in exit_status are in French, therefore ask R to look for "Décédé" instead of "Dead", "DOA", or "DD".

```{r}
## use this version for coded values (e.g. "DOA", "DD")
linelist_cleaned$DIED <- linelist_cleaned$exit_status %in% c("Décédé")
```


### Creating new values for exit status
```{r}
# Tabulate to see all possible values of exit_status
table(linelist_cleaned$exit_status, useNA = "always")

linelist_cleaned <- linelist_cleaned %>% 
  mutate(exit_status2 = case_when(
    exit_status == "Décédé"          ~ "Died",
    exit_status == "-"               ~ NA_character_,
    exit_status == "Echappé"         ~ "Left",
    exit_status == "Déchargé/Guéri"  ~ "Discharged"
  ))

# Table the NEW variable to check transformation
table(linelist_cleaned$exit_status2, useNA = "always")
```


### Create case definition variable

```{r}
# Create a factor variable based on rules from other simple character variables
# If you have access to lab results, you can create a case definition variable 
# the tilda (~) is used to assign the new values (Conf, prob, susp, unknown)
# starting from the specific to the general
# TRUE assigns all remaining rows 
# You MUST modify this section to match your case definition. The below
# uses positive RDT for Confirmed and epi link only for Probable.
#
linelist_cleaned <- linelist_cleaned %>%
  mutate(case_def = case_when(
    is.na(hep_e_rdt) & is.na(other_cases_in_hh)           ~ NA_character_,
    hep_e_rdt == "Positive"                               ~ "Confirmed",
    hep_e_rdt != "Positive" & other_cases_in_hh == "Yes"  ~ "Probable",
    TRUE                                                  ~ "Suspected"
  ))
```

Create list of symptom variables - **already switched "med" variables to these ones, but some do not apply and must be removed from the list, such as convulsions and history of fever?**

```{r}
# vectors of variable names ----------------------------------------------------

## You may want to group the names of several variables that have the same possible 
## values in to a named vector. 
## This way if you want to run the same function over these variables you can 
## simply use the named vector rather than typing out each variable individually

# create a grouping of all symptoms 
SYMPTOMS <- c("generalized_itch", 
              #"history_of_fever", 
              "fever",
              "joint_pains",
              "epigastric_pain_heartburn",
              "nausea_anorexia",
              "vomiting", 
              "diarrhoea",
              "bleeding", 
              "headache",
              "mental_state",
              #"convulsions",
              "other_symptoms"  
              )

# create a grouping of all lab tests 
LABS <- c("hep_b_rdt", 
          "hep_c_rdt",
          "hep_e_rdt",
          "test_hepatitis_a",
          "test_hepatitis_b",
          "test_hepatitis_c",
#          "test_hepatitis_e_igg",
          "test_hepatitis_e_igm" ,
          "test_hepatitis_e_genotype",
          "test_hepatitis_e_virus",
          "malaria_rdt_at_admission",
#          "malaria_blood_film", 
          "dengue",
#          "dengue_rdt", 
          "yellow_fever",
#          "typhoid",
          "chikungunya_onyongnyong", 
#          "ebola_marburg",
#          "lassa_fever",
          "other_arthropod_transmitted_virus" 
#          "other_pathogen"
          )

```

### **Drop observations**  

**Removing those with dates of onset outside of years 2016 or 2017**

```{r}
# Check date range of date_of_onset
range(linelist_cleaned$date_of_onset, na.rm = T)

# Filter to keep observations where date_of_onset after April 2016 AND before October 2017 (OR missing)
linelist_cleaned <- linelist_cleaned %>%
  filter((date_of_onset >= as.Date("2016-04-01") & date_of_onset <= as.Date("2017-10-01")) | is.na(date_of_onset))
# Re-check the date range
range(linelist_cleaned$date_of_onset, na.rm = T)
```



## Export the cleaned dataset

```{r save_cleaned_data, eval = FALSE}
## OPTIONAL: save your cleaned dataset as an excel file! 
## put the current date in the name so you know!
writexl::write_xlsx(linelist_cleaned, paste0("linelist_cleaned_", Sys.Date(), ".xlsx"))
```


Remmber, you defined reporting_week at beginning of script
```{r}

# drop cases after current week
# you will probably have cases in your dataset if case finding continues
# as these are after the week of reporting, drop them
linelist_cleaned <- linelist_cleaned %>% 
  filter(date_of_onset <= week2date(sprintf("%s-7", reporting_week)))

# define the first week of outbreak (date of first case)
first_week <- levels(linelist_cleaned$epiweek)[1]

# outbreak start 
# return the first day in the week of first case 
obs_start <- week2date(sprintf("%s-1", first_week))

# return last day of reporting week 
obs_end   <- week2date(sprintf("%s-7", reporting_week))

```





Demographic Tables:

These work:

Cases by age group and definition 

```{r describe_by_age_group_and_def}
# use if you have lab results in your data
# get counts and props of age groups by case definition 
# include column and row totals 
descriptive(linelist_cleaned, "age_group", "case_def", coltotals = TRUE, rowtotals = TRUE) %>%
  rename("Age group" = age_group) %>%
  rename_redundant(prop = "%") %>%
  augment_redundant("_n$" = " cases (n)") %>%
  kable(digits = 2)
```




Cases by age group and sex 

```{r describe_by_age_group_and_sex}
descriptive(linelist_cleaned, "age_group", "sex", coltotals = TRUE, rowtotals = TRUE) %>% 
  rename("Age group" = age_group) %>%
  rename_redundant(prop = "%") %>%
  augment_redundant("_n$" = " cases (n)") %>%
  kable(digits = 2)
```


Alternatively if you would like proportions to be of the total population, use the following. 

```{r total_props_agegroup_sex}
descriptive(linelist_cleaned, "age_group", "sex", 
            coltotals = TRUE, rowtotals = TRUE, proptotal = TRUE) %>% 
  rename("Age group" = age_group) %>%
  rename_redundant(prop = "%") %>%
  augment_redundant("_n$" = " cases (n)") %>%
  kable(digits = 2)
```


Age pyramid 

There were `r fmt_count(linelist_cleaned, is.na(sex))` cases missing information on sex and 
`r fmt_count(linelist_cleaned, is.na(age_group))` missing age group.

```{r age_pyramid, warning=FALSE}
# plot age pyramid 
 
plot_age_pyramid(linelist_cleaned, 
                 age_group = "age_group", 
                 split_by = "sex") + 
  labs(y = "Cases (n)", x = "Age group") + # change axis  labels (nb. x/y flip)
  theme(legend.position = "bottom",     # move legend to bottom
        legend.title = element_blank(), # remove title
        text = element_text(size = 18)  # change text size
       )
```

Age pyramid for Under 2 obviously doesn't work.





```{r describe_by_symptoms}

# get counts and proportions for all variables named in SYMPTOMS
multi_descriptive(linelist_cleaned, SYMPTOMS) %>% 
  # re-order columns (drop total_prop too)
  select(symptom, Yes_n, Yes_prop, No_n, No_prop, Total_n) %>%
  # rename accordingly
  rename_redundant("prop" = "%") %>%
  augment_redundant("_n$" = " (n)") %>%
  kable(digits = 2)


```

```{r describe_by_labs}
# get counts and proportions for all variables named in LABS
multi_descriptive(linelist_cleaned, LABS, .id = "Lab test") %>% 
  # re-order columns (drop total_prop too)
  select("Lab test", Pos_n, Positive_prop, Neg_n, Negative_prop, 
         "Not done_n", "Not done_prop", Total_n) %>% 
  # rename accordingly
  rename("Unknown (n)" = "Not done_n") %>%
  rename_redundant("prop" = "%") %>%
  augment_redundant("_n$" = " (n)") %>%
  kable(digits = 2)

```

CFR Section
Skip time-to-death because we don't have that variable.

Case Fatality Ratio - note that this requires there to be a variable patient_facilit_type, the value "Inpatient", and the DIED variable created earlier. 

```{r overall_cfr, warning = FALSE, message = FALSE}
# use arguments from above to produce overal CFR
overall_cfr <- linelist_cleaned %>% 
  filter(patient_facility_type == "Inpatient") %>%
  case_fatality_rate_df(deaths = DIED, mergeCI = TRUE) %>%
  rename("Deaths" = deaths,
         "Cases" = population,
         "CFR (%)" = cfr,
         "95%CI" = ci)

  knitr::kable(overall_cfr, digits = 2)         # print nicely with 2 digits
```



We can also do CFR by sex

```{r cfr_by_sex, warning = FALSE, message = FALSE}
linelist_cleaned %>%
  filter(patient_facility_type == "Inpatient") %>%
  mutate(sex = forcats::fct_explicit_na(sex, "-")) %>%
  case_fatality_rate_df(deaths = DIED, group = sex, mergeCI = TRUE, add_total = TRUE) %>%
  rename("Sex" = sex, 
         "Deaths" = deaths, 
         "Cases" = population, 
         "CFR (%)" = cfr, 
         "95%CI" = ci) %>% 
  knitr::kable(digits = 2)
```


...and by age_group:


CFR by age group among inpatients with known outcomes

```{r cfr_by_age_group, warning = FALSE, message = FALSE}

linelist_cleaned %>%
  filter(patient_facility_type == "Inpatient") %>%
  case_fatality_rate_df(deaths = DIED, group = age_group, mergeCI = TRUE, add_total = TRUE) %>%
  tidyr::complete(age_group, 
                  fill = list(deaths = 0, population = 0, cfr = 0, ci = 0)) %>% # Ensure all levels are represented
  rename("Age Group" = age_group, 
         "Deaths" = deaths, 
         "Cases" = population, 
         "CFR (%)" = cfr, 
         "95%CI" = ci) %>% 
  knitr::kable(digits = 2)
```


CFR by case definition (not enough to do)

```{r cfr_by_case_def, warning = FALSE, message = FALSE}
# Use if you have enough confirmed cases for comparative analysis 
#
 linelist_cleaned %>%
   filter(patient_facility_type == "Inpatient") %>%
   case_fatality_rate_df(deaths = DIED, group = case_def, mergeCI = TRUE, add_total = TRUE) %>%
   rename("Case Definition" = case_def, 
          "Deaths" = deaths, 
          "Cases" = population, 
          "CFR (%)" = cfr, 
          "95%CI" = ci) %>% 
   knitr::kable(digits = 2)
```


## Attack Rate
To use the attack rate section, we need to modify the first code slightly. An object "population" is created from the sum of population counts in the population figures. Because we only imported region-based population counts, we must change this command to reflect that we do not have population_data_age, but rather population_data_region

```{r eval=FALSE}
# OLD define population 
# population <- sum(population_data_age$population)
```

Running the correct command and printing the value of population, we see that the sum population across regions is estimated to be 62336.
```{r}
# Correct command for this exercise
population <- sum(population_data_region$population)

population
```

Below gives the attack rate per 10,000 population (N = `r format(population, big.mark = ",")`)
```{r attack_rate}
# calculate the ar
# store as AR to be able to use output for automating text below (inline functions)
ar <- attack_rate(nrow(linelist_cleaned), population, multiplier = 10000)

ar %>%
  merge_ci_df(e = 3) %>% # merge the lower and upper CI into one column
  rename("Cases (n)" = cases, 
         "Population" = population, 
         "AR (per 100,000)" = ar, 
         "95%CI" = ci) %>% 
  select(-Population) %>% # drop the population column as it is not changing
  knitr::kable(digits = 2, align = "r")
```

We are unable to calculate the attack rate by age group, because we do not have population counts for each age group.

Mortality attributable to AJS - not appropriate for this example because...




## Time

```{r create_incidence, message = FALSE}
# This code creates case counts for each week of your outbreak, overall
# As with aweek, you can change the start of your week to e.g. "Sunday week"
inc_week_7 <- incidence(linelist_cleaned$date_of_onset, interval = "Monday week")

# this sets the theme in ggplot for epicurves
epicurve_theme <- theme(
  axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1), 
  legend.title = element_blank(),
  panel.grid.major.x = element_line(color = "grey60", linetype = 3),
  panel.grid.major.y = element_line(color = "grey60", linetype = 3)
)
# This sets the labels in ggplot for the epicurves
epicurve_labels <- labs(x = "Calendar week", 
                        y = "Cases (n)", 
                        title = "Cases by week of onset",
                        subtitle = sprintf("Source: MSF data from %s", reporting_week)
                       ) 
```


```{r epicurve, message = FALSE}
# plot your epicurve as a ggplot (incidence::plot is different to base::plot)
basic_curve <- plot(inc_week_7, show_cases = TRUE, border = "black", n_breaks = nrow(inc_week_7)) + 
  scale_y_continuous(expand = c(0, 0)) +  # set origin for axes
  # add labels to axes and below chart
  epicurve_labels +
  # change visuals of dates and remove legend title
  epicurve_theme

# show your plot (stored for later use) 
basic_curve
basic_curve + scale_x_incidence(inc_week_7, n_breaks = 6)
```


Other possible variants on the epi curve:  
* Option 1  
* Option 2  
* Option 3  


The below gives the attack rate per week. 

```{r attack_rate_per_week, warning = FALSE, message = FALSE}
# counts and cumulative counts by week
cases <- linelist_cleaned %>%
  arrange(date_of_onset) %>%        # arrange by date of onset
  count(epiweek, .drop = FALSE) %>% # count all epiweeks and include zero counts
  mutate(cumulative = cumsum(n))    # add a cumulative sum

# attack rate for each week
ar <- attack_rate(cases$n, population, multiplier = 100000) %>% 
  bind_cols(select(cases, epiweek), .) # add the epiweek column to table

ar %>%
  merge_ci_df(e = 4) %>% # merge the lower and upper CI into one column
  rename("Epiweek" = epiweek, 
         "Cases (n)" = cases, 
         "Population" = population, 
         "AR (per 100,000)" = ar, 
         "95%CI" = ci) %>% 
  knitr::kable(digits = 2, align = "r")
```

And cumulatively:  


```{r cumulative_attack_rate_per_week}
# cumulative attack rate by week
attack_rate(cases$cumulative, population, multiplier = 100000) %>% 
  bind_cols(select(cases, epiweek), .) %>% # add the epiweek column to table
  merge_ci_df(e = 4) %>% # merge the lower and upper CI into one column
  rename("Epiweek" = epiweek, 
         "Cases (n)" = cases, 
         "Population" = population, 
         "AR (per 100,000)" = ar, 
         "95%CI" = ci) %>% 
  knitr::kable(digits = 2, align = "r")
```

```{r cfr_per_week, warning = FALSE, message = FALSE}
# group by known outcome and case definition 
cfr <- linelist_cleaned %>%
  filter(patient_facility_type == "Inpatient") %>%
  case_fatality_rate_df(grepl("Dead", exit_status), group = epiweek)

cfr %>%
  merge_ci_df(e = 4) %>% # merge the lower and upper CI into one column
  rename("Epiweek" = epiweek, 
         "Deaths" = deaths, 
         "Cases" = population, 
         "CFR (%)" = cfr, 
         "95%CI" = ci) %>% 
  knitr::kable(digits = 2, align = "r")
```

You could plot the AR (in the population) and CFR (among inpatients only) together as line graphs by epiweek. 

```{r ar_line_graph}
ar_plot <- ggplot(ar, aes(x = week2date(epiweek) + (7 * 0.5), group = 1)) +
  geom_ribbon(aes(ymin = lower, ymax = upper), 
              color = "blue", fill = "blue", linetype = 2, alpha = 0.2, show.legend = FALSE) +
  geom_line(aes(y = ar), color = "blue", show.legend = FALSE) +
  scale_y_continuous(expand = c(0, 0)) +  # set origin for axes
  # scale the x axis the same as the incidence curve. Expand forces it to align. 
  incidence::scale_x_incidence(inc_week_7, n_breaks = nrow(inc_week_7), expand = c(0, 7 * 1.5)) +
  # add labels to axes and below chart
  labs(x = "Calendar week", y = "AR [95% CI]", subtitle = "Attack Rate (per 100,000)") + 
  # change visuals of dates and remove legend title
  epicurve_theme
```

To plot CFR you must first make the CFR table using above code

```{r cfr_line_graph}
cfr_plot <- ggplot(cfr, aes(x = week2date(epiweek) + (7 * 0.5), group = 1)) +
  geom_ribbon(aes(ymin = lower, ymax = upper), 
              color = "red", fill = "red", linetype = 2, alpha = 0.2, show.legend = FALSE) +
  geom_line(aes(y = cfr), color = "red", show.legend = FALSE) +
  scale_y_continuous(expand = c(0, 0)) +  # set origin for axes
  # scale the x axis the same as the incidence curve. Expand forces it to align. 
  incidence::scale_x_incidence(inc_week_7, n_breaks = nrow(inc_week_7), expand = c(0, 7 * 1.5)) +
  # add labels to axes and below chart
  labs(x = "Calendar week", y = "CFR [95% CI]", 
       subtitle = "Case Fatality Ratio [95% CI] Among Inpatients") + 
  # change visuals of dates and remove legend title
  epicurve_theme 

```


You could then also add the AR and CFR by week on to an epicurve. 

```{r epicurve_ar_cfr, message = FALSE, fig.height = 10}
nofx <- theme(axis.text.x = element_blank(),
              axis.title.x = element_blank())
cowplot::plot_grid(
  basic_curve + nofx,
  ar_plot + nofx,
  cfr_plot,
  align = "v", # align plots vertically
  axis = "lr", # only by their left and right margins
  ncol = 1     # allow only one column
)
```


more optins...



## Place


Basic descriptive table:

```{r describe_by_region_facility}
# get counts and props of region by facility 
# include column and row totals 

descriptive(linelist_cleaned, "patient_origin", 
            "patient_facility_type",
            coltotals = TRUE, rowtotals = TRUE) %>% 
  rename("Region" = "patient_origin") %>%
  augment_redundant("_n$" = " (n)") %>% # modify _n to (n)
  rename_redundant("prop" = "%")    %>% # rename proportions to %
  kable(digits = 2)
```


And attack rates by region.   
Note: changed population_data_region to population_data in line 2.
Error received on left_join: Error in tbl_vars(y) : object 'population_data_region' not found

Then, Error received from merge to create cases. Turns out that in population counts the regions are ALL CAPITALS but in linelist the are lowercase. Need to add line to make that solved.

```{r}
# Converts patient_origin variable to uppercase, to match with population_data
linelist_cleaned$patient_origin <- toupper(linelist_cleaned$patient_origin)
```

Also need fix in proportion() function in sitrep to is.na(n)

```{r attack_rate_by_region}

## - [ ] consider facet wrapping by an overarching unit if have many regions (e.g. by province)


cases <- count(linelist_cleaned, patient_origin) %>%   # cases for each region
  left_join(population_data, by = "patient_origin")    # merge population data 
# attack rate for region
ar <- attack_rate(cases$n, cases$population, multiplier = 100) %>% 
  # add the region column to table
  bind_cols(select(cases, patient_origin), .) %>% 
  rename("Region" = patient_origin, 
         "Cases (n)" = cases, 
         "Population" = population, 
         "AR (per 100,000)" = ar, 
         "Lower 95%CI" = lower,
         "Upper 95%CI" = upper) 

ar %>% 
  merge_ci_df(e = 4) %>% # merge lower and upper CI in to one column 
  rename("95%CI" = ci) %>%  # rename single 95%CI column
  kable(digits = 2, align = "r", format.args = list(big.mark = ",")) # set thousands separator
```


You could then also plot this on a bar chart with confidence intervals. 

```{r bar_attack_rate_by_region}
# plot with the region on the x axis sorted by increasing ar
# ar value on the y axis 
ggplot(ar, aes(x = reorder(Region, `AR (per 100,000)`),
               y = `AR (per 100,000)`)) + 
  geom_bar(stat = "identity", col = "black", fill = "red") + # plot as bars (identity = as is)
  geom_errorbar(aes(ymin = `Lower 95%CI`, ymax = `Upper 95%CI`), width = 0.2) + # add CIs
  scale_y_continuous(expand = c(0,0)) +  # set origin for axes
  # add labels to axes and below chart
  labs(x = "Region", y = "AR (per 100,000)", 
       captions = paste0("Source: MSF data from ", reporting_week)) + 
  epicurve_theme
```



Mortality Rates by region
Changed population_data_region to population_data

Changed "Dead" to "Décédé" in grepl


```{r mortality_rate_region}

deaths <- group_by(linelist_cleaned, patient_origin) %>%
  filter(grepl("Décédé", exit_status)) %>% 
  summarise(deaths = n()) %>% # count deaths by region
  left_join(population_data, by = "patient_origin") # merge population data 

mortality_rate(deaths$deaths, deaths$population, multiplier = 10000) %>%
  # add the region column to table
  bind_cols(select(deaths, patient_origin), .) %>% 
  merge_ci_df(e = 4) %>% # merge the lower and upper CI into one column
  rename("Region" = patient_origin, 
         "Deaths" = deaths, 
         "Population" = population, 
         "Mortality (per 10,000)" = `mortality per 10 000`, 
         "95%CI" = ci) %>% 
  kable(digits = 2)
```
